src/lsm/forest.zig

```diff
@@ -425,40 +437,50 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
             // last beat.
             // The first bar has no manifest compaction.
             if ((last_beat or last_half_beat) and op > constants.lsm_compaction_ops) {
-                forest.manifest_log_progress = .compacting;
                 forest.manifest_log.compact(compact_manifest_log_callback, op);
-                forest.compaction_progress = .trees_and_manifest;
             } else {
-                assert(forest.manifest_log_progress == .idle);
+                forest.compaction_progress.?.manifest_log_done = true;
             }
         }
 
-        fn compact_callback(forest: *Forest) void {
+        fn compact_trees_callback(forest: *Forest) void {
             assert(forest.progress.? == .compact);
-            assert(forest.compaction_progress != .idle);
+            assert(forest.compaction_progress != null);
+            assert(!forest.compaction_progress.?.trees_done);
+            forest.compaction_progress.?.trees_done = true;
 
-            if (forest.compaction_progress == .trees_and_manifest) {
-                assert(forest.manifest_log_progress != .idle);
+            if (forest.compaction_progress.?.all_done()) {
+                forest.compact_finish();
             }
+        }
 
-            forest.compaction_progress = if (forest.compaction_progress == .trees_and_manifest)
-                .trees_or_manifest
-            else
-                .idle;
-            if (forest.compaction_progress != .idle) return;
+        fn compact_manifest_log_callback(manifest_log: *ManifestLog) void {
+            const forest: *Forest = @fieldParentPtr("manifest_log", manifest_log);
 
-            forest.verify_table_extents();
+            assert(forest.progress.? == .compact);
+            assert(forest.compaction_progress != null);
+            assert(!forest.compaction_progress.?.manifest_log_done);
+            forest.compaction_progress.?.manifest_log_done = true;
+
+            if (forest.compaction_progress.?.all_done()) {
+                forest.compact_finish();
+            }
+        }
+
+        fn compact_finish(forest: *Forest) void {
+            assert(forest.progress.? == .compact);
+            assert(forest.compaction_progress != null);
+            assert(forest.compaction_progress.?.trees_done == true);
+            assert(forest.compaction_progress.?.manifest_log_done == true);
```

Comment:
Nit, but this could just be
```suggestion
            assert(forest.compaction_progress.?.trees_done);
            assert(forest.compaction_progress.?.manifest_log_done);
```